### ISSUE TITLE
GPU environment is not working properly with google drive.

### ISSUE THREAD

**Issue by vikalpjain91** [googlecolab/colabtools#1227]
Bug report for Colab: http://colab.research.google.com/.

For questions about colab usage, please use [stackoverflow](https://stackoverflow.com/questions/tagged/google-colaboratory/).

- Describe the current behavior:
Colab is giving Input/Output error intermittently when we switch the colab to GPU. It is working fine when we use standard environment.

- Describe the expected behavior:
It should not show any Input/Output Error.

- The web browser you are using (Chrome, Firefox, Safari, etc.):

- Link to a self-contained notebook that reproduces this issue
  (click the *Share* button, then *Get Shareable Link*):
https://colab.research.google.com/drive/1o1A9WpENxEtY6CxlHBQrga9QJvj-JExX?usp=sharing

In my notebook there are lots of image chunks are available it is approximately 20k+20k+9k+9k in 4 directories which are not in the main directory. whenever I am using the standard environment it is working like a charm. though sometimes the GPU environment is also working fine which is sometimes rare.
![image](https://user-images.githubusercontent.com/16263750/81454133-08b80680-91a9-11ea-957c-a2ce116e1680.png)

After some time I have tried which was working fine
![image](https://user-images.githubusercontent.com/16263750/81454319-87ad3f00-91a9-11ea-8aef-242c27d18768.png)

Apart from that notebook is giving one more issue of not running the GPU in the initial phase of training the model. wondering if it is linked with the same drive issue as it has to pick up data from the drive.
![image](https://user-images.githubusercontent.com/16263750/81454481-1457fd00-91aa-11ea-9528-e2b1cde0aae1.png)

generally for the first few epochs, it will take 2hr 30 minutes. which is limiting me to use the GPU efficiently once the initial few epoch is over it is working like a charm and a complete epoch can be completed in a few seconds(50 sec).
Just to give an idea how much time it is taking currently to run few steps:
observe time here:
![image](https://user-images.githubusercontent.com/16263750/81454627-67ca4b00-91aa-11ea-9974-e0c77c6b82a9.png)

Observe time here again:
![image](https://user-images.githubusercontent.com/16263750/81454755-cd1e3c00-91aa-11ea-89f1-f332f8193a87.png)

The machine is in the GPU environment.
![image](https://user-images.githubusercontent.com/16263750/81454812-e921dd80-91aa-11ea-8256-17cc879112ae.png)

Notebook link: https://colab.research.google.com/drive/1o1A9WpENxEtY6CxlHBQrga9QJvj-JExX?usp=sharing

Though we have to connect it to my drive to reproduce the issue.
**Any workaround can help me a ton because I am using colab for my academic assignments.**
Unfortunately, I can not buy Pro from India. if there is a workaround for that too happy to do so if it resolves my problem. Seriously in need of colab right now.

**Comment by colaboratory-team** [comment]
58k files in 4 drive folders will not work well. The fact that it works more in CPU VMs than GPU VMs and intermittently works in GPU VMs is an implementation coincidence.
The most robust workaround to this problem would be to archive your 58k files into a single or a few files (e.g. using zip or tar) and copying them from Drive to the VM and extracting there, before beginning your ML workload.
This can only work if your dataset is small enough to fit on the VM's attached disk, of course.
If your dataset is too large to do that and you cannot usefully split it up, your only alternative while using Drive will be to split up your 4 folders into many more.

**Comment by vikalpjain91** [comment]
Thanks, @colaboratory-team . One more thing that I am talking about the GPU was not working efficiently for the first few epochs! Is it related to the Google Drive issue? Or there is any other workaround Or I have to raise a new bug for that?

**Comment by colaboratory-team** [comment]
Yes if your data needs to get paged in from Drive a slow start would be expected.

**Comment by vikalpjain91** [comment]
Thanks

**Comment by vikalpjain91** [comment]
Sorry But I am unable to find any reference to do copy into VM's local disk. Can you please help me out if Colab have shared any references that you were talking about. I want to implement it really hard.
@colaboratory-team ?

**Comment by colaboratory-team** [comment]
You're already mounting your Google Drive on the VM so to unzip an archive from there to the VM's local disk you can simply extract it anywhere outside the Drive mount, e.g.
```
!unzip '/content/drive/My Drive/path/to/data.zip' -d /content/localdata
```
and then use the data from `/content/localdata` (or whatever path you use after `-d` in the above command). This of course assumes you've placed your data in a zip file at the path named (modify the command to point to wherever you've placed the .zip).

**Comment by vikalpjain91** [comment]
I am copying the files into the location what you have mentioned but things are not working as you said.

https://colab.research.google.com/drive/1UfGIQqh900vqBDspMXxotJs_BrcGlbdE?usp=sharing

it is more or less like resource got exhausted which was not the case earlier.

```
'''Moving out files to vm'''
import os
VM_LOCATION = "Vikalp"
VM_PICKLES = os.path.join(LOCATION, "pickles")
VM_TRAINING_LOCATION = os.path.join(LOCATION, "training")
VM_FAKE_LOCATION = os.path.join(TRAINING_LOCATION, "fake")
VM_GROUND_TRUTH_LOCATION = os.path.join(LOCATION, "ground_truth")
VM_TESTING_GROUND_TRUTH_LOCATION = os.path.join(LOCATION, "testing_ground_truth")
VM_PRISTINE_LOCATION = os.path.join(TRAINING_LOCATION, "pristine")
VM_TESTING_LOCATION = os.path.join(LOCATION, "testing")
VM_TESTING_FAKE_LOCATION = os.path.join(TESTING_LOCATION, "fake")
VM_TESTING_PRISTINE_LOCATION = os.path.join(TESTING_LOCATION, "pristine")

if not os.path.exists(VM_LOCATION):
    print("Creating directory")
    os.mkdir(VM_LOCATION)
	
if not os.path.exists(VM_TRAINING_LOCATION):
    print("Creating directory")
    os.mkdir(VM_TRAINING_LOCATION)
	
if not os.path.exists(VM_TESTING_LOCATION):
    print("Creating directory")
    os.mkdir(VM_TESTING_LOCATION)
	
if not os.path.exists(VM_GROUND_TRUTH_LOCATION):
    print("Creating directory")
    os.mkdir(VM_GROUND_TRUTH_LOCATION)

if not os.path.exists(VM_TESTING_GROUND_TRUTH_LOCATION):
    print("Creating directory")
    os.mkdir(VM_TESTING_GROUND_TRUTH_LOCATION)

if not os.path.exists(VM_TESTING_LOCATION):
    print("Creating directory")
    os.mkdir(VM_TESTING_LOCATION)

if not os.path.exists(VM_TESTING_FAKE_LOCATION):
    print("Creating directory")
    os.mkdir(VM_TESTING_FAKE_LOCATION)

if not os.path.exists(VM_TESTING_PRISTINE_LOCATION):
    print("Creating directory")
    os.mkdir(VM_TESTING_PRISTINE_LOCATION)

VM_BIN_GROUND_TRUTH_LOCATION = os.path.join(VM_LOCATION, "binary_ground_truth")
VM_TEST_BIN_GROUND_TRUTH_LOCATION = os.path.join(VM_LOCATION, "test_binary_ground_truth")

if not os.path.exists(VM_BIN_GROUND_TRUTH_LOCATION):
    print("Creating directory")
    os.mkdir(VM_BIN_GROUND_TRUTH_LOCATION)
	
if not os.path.exists(VM_TEST_BIN_GROUND_TRUTH_LOCATION):
    print("Creating directory")
    os.mkdir(VM_TEST_BIN_GROUND_TRUTH_LOCATION)


import shutil

def move_files(SOURCE, DEST):
  SOURCE_FILES = os.listdir(SOURCE)
  print("Total files to copy: ", len(SOURCE_FILES))
  for fl in SOURCE_FILES:
    if not os.path.exists( os.path.join(DEST, fl) ):
      shutil.copy(os.path.join(SOURCE, fl), os.path.join(DEST, fl))
    print("", end=".")
  print()
  print("Total files in destination: ", len(os.listdir(DEST)))

move_files(FAKE_LOCATION, VM_FAKE_LOCATION)
move_files(PRISTINE_LOCATION, VM_PRISTINE_LOCATION)
move_files(TESTING_FAKE_LOCATION, VM_TESTING_FAKE_LOCATION)
move_files(TESTING_PRISTINE_LOCATION, VM_TESTING_PRISTINE_LOCATION)

Creating a data generator same as above.
```
'''commenting this section as colab is not supporting larger number of chunks from the drive'''
import keras
from tensorflow.keras.preprocessing.image import ImageDataGenerator
import matplotlib.pyplot as plt
from keras import metrics
KERNEL_SIZE  =256
epochs = 15
batch_size=128
# total_train = len(os.listdir(FAKE_VALIDATION_PF_CHUNKS_LOCATION))+len(os.listdir(FAKE_TRAIN_PF_CHUNKS_LOCATION))
# total_validation = len(os.listdir(PRISTINE_VALIDATION_PF_CHUNKS_LOCATION))+len(os.listdir(PRISTINE_TRAIN_PF_CHUNKS_LOCATION))
# total_train = 20629+20630 #IO error from google drive
# total_validation = 9371+9372 #IO error from google drive

train_image_generator = ImageDataGenerator(rescale=1./255,
        shear_range=0.2,
        zoom_range=0.2,
        horizontal_flip=True)

validation_image_generator = ImageDataGenerator(rescale=1./255)


train_data_gen = train_image_generator.flow_from_directory(batch_size=batch_size,
                                                           directory=VM_TRAINING_LOCATION,
                                                           shuffle=True,
                                                           target_size=(KERNEL_SIZE, KERNEL_SIZE),
                                                           class_mode='binary')


val_data_gen = validation_image_generator.flow_from_directory(batch_size=batch_size,
                                                             directory=VM_TESTING_LOCATION,
                                                              target_size=(KERNEL_SIZE, KERNEL_SIZE),
                                                              class_mode='binary')

```


gives me the error of resource exhaust at the time of training.
```
epochs = 300
history = model_aug.fit_generator(
    train_data_gen,
    steps_per_epoch=100,
    epochs=epochs, 
    # validation_steps=100,
    callbacks=callbacks,
    validation_data=val_data_gen, 
    )
# Plot training & validation loss values
plt.plot(history.history['loss'])
plt.plot(history.history['val_loss'])
plt.title('Model loss')
plt.ylabel('Loss')
plt.xlabel('Epoch')
plt.legend(['Train', 'Test'], loc='upper left')
plt.show()
```

Error:
![image](https://user-images.githubusercontent.com/16263750/81632439-ea137300-9427-11ea-9b54-bdc273cf9693.png)


```

**Comment by Santosh-Gupta** [comment]
> You're already mounting your Google Drive on the VM so to unzip an archive from there to the VM's local disk you can simply extract it anywhere outside the Drive mount, e.g.
> 
> ```
> !unzip '/content/drive/My Drive/path/to/data.zip' -d /content/localdata
> ```
> 
> and then use the data from `/content/localdata` (or whatever path you use after `-d` in the above command). This of course assumes you've placed your data in a zip file at the path named (modify the command to point to wherever you've placed the .zip).

Is there a way for non-zipped files. I am working with a 24 gb numpy memmap, and I can maybe download it once to my local without any problem, but after that I run into some sort of rate limit transfer exceeded issue. 

https://stackoverflow.com/questions/61842823/is-there-a-way-to-copy-a-google-drive-file-into-colab-without-using-some-sort-o
